sled-agent/src/rack_setup/service.rs

```diff
@@ -0,0 +1,227 @@
+                let client = sled_agent_client::Client::new_with_client(
+                    &format!("http://{}", request.sled_address),
+                    client,
+                    self.log.new(o!("SledAgentClient" => request.sled_address)),
+                );
+
+                info!(self.log, "sending partition requests...");
+                for partition in &request.partitions {
+                    let filesystem_put = || async {
+                        info!(self.log, "creating new filesystem: {:?}", partition);
+                        client.filesystem_put(&partition.clone().into())
+                            .await
+                            .map_err(BackoffError::transient)?;
+                        Ok::<
+                            (),
+                            BackoffError<
+                                sled_agent_client::Error<sled_agent_client::types::Error>,
+                            >,
+                        >(())
+                    };
+                    let log_failure = |error, _| {
+                        warn!(self.log, "failed to create filesystem"; "error" => ?error);
+                    };
+                    retry_notify(
+                        internal_service_policy(),
+                        filesystem_put,
+                        log_failure,
+                    ).await?;
+                }
+                Ok(())
+            })
+        ).await.into_iter().collect::<Result<Vec<()>, SetupServiceError>>()?;
+
+        // Issue service initialization requests.
+        //
+        // Note that this must happen *after* the partition initialization,
+        // to ensure that CockroachDB has been initialized before Nexus
+        // starts.
+        futures::future::join_all(
+            config.requests.iter().map(|request| async move {
+                info!(self.log, "observing request: {:#?}", request);
+                let dur = std::time::Duration::from_secs(60);
+                let client = reqwest::ClientBuilder::new()
+                    .connect_timeout(dur)
+                    .timeout(dur)
+                    .build()
+                    .map_err(|e| nexus_client::Error::<nexus_client::types::Error>::from(e))?;
+                let client = sled_agent_client::Client::new_with_client(
+                    &format!("http://{}", request.sled_address),
+                    client,
+                    self.log.new(o!("SledAgentClient" => request.sled_address)),
+                );
+
+                info!(self.log, "sending service requests...");
+                let services_put = || async {
+                    info!(self.log, "initializing sled services: {:?}", request.services);
+                    client.services_put(
+                        &sled_agent_client::types::ServiceEnsureBody {
+                            services: request.services.iter().map(|s| s.clone().into()).collect()
+                        })
```

Comment:
Another clarification question, not totally related to this PR, but what's the distinction between a "filesystem" request and a "services" request here? Both seem to result in an SMF service, launched in a zone, listening on some address. Is the creation of the ZFS dataset for the service's data what distinguishes these?

Reply:
- "Filesystem requests" request a portion of the underlying disk space. They create a new ZFS filesystem which must be co-located with the service. Example: Crucible, Cockroach, Clickhouse.
- "Service requests" request services that are decoupled from the underlying sled. They should not rely on any local storage. Example: Nexus, Oximeter.

I could lump the requests together, I suppose? Implementation detail: They're handled by different portions of the Sled Agent, since storage requires more explicit handling.